### PR TITLE
Move activities and places into their own modules.

### PR DESCRIPTION
**Pull request by maffoo**
Pull the Activity classes out into a separate module instead of putting them inside an event callback in the main app module. Among other things, this forces us to declare all the dependencies for each activity in its constructor, rather than close over variables from the outer scope when creating activities. We also put the functions for generating urls into a single location so they can be more self-contained.

### CODE REVIEW

**Comment by joshm-3b6b58** [comment]
Sorry, I've been in the cleanroom most of the week, I'll take a look at this soon :/

**Comment by joshm-3b6b58** [comment]
Some comments about the overall structure of this would be very helpful for people trying to navigate this to add functionality.

File: client-js/app/scripts/activities.ts [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+    var breadcrumbs = [];
+    for (var i = 0; i <= this.path.length; i++) {
+      breadcrumbs.push({
+        name: (i == 0) ? 'grapher' : this.path[i-1],
+        isLink: i < this.path.length,
+        url: places.grapherUrl(this.path.slice(0, i))
+      });
+    }
+    var breadcrumbExtras = [
```
Comment (joshm-3b6b58) [inline client-js/app/scripts/activities.ts:84]: This can be broken out to be more readable as on line 87

File: client-js/app/scripts/app.ts [hunk 2]
```diff
@@ -2,10 +2,10 @@
 
 import page from "page";
 
-import {Activity, ActivityState} from "./activity";
-import {AsyncQueue} from './async-queue';
-import {Lifetime} from "./lifetime";
+import {Activity} from "./activity";
+import * as activities from "./activities";
 import * as manager from "./manager";
+import * as places from "./places";
```
Comment (joshm-3b6b58) [inline client-js/app/scripts/app.ts]: Nit, this import isn't used.

File: client-js/app/scripts/places.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+// Function for creating URLs of all the places in the application.
+
```
Comment (joshm-3b6b58) [inline client-js/app/scripts/places.ts:2]: So this just creates URLs, it doesn't handle routing at all?

**Comment by joshm-3b6b58** [comment]
Other than I think it needs some commentary it LGTM. For example, if I want to add a widget to the web app, when do I create an activity? Why do I need a place? What goes is app.ts?

File: client-js/app/scripts/activities.ts (hunk 1 above)
Comment (maffoo) [inline client-js/app/scripts/activities.ts:84]: done

File: client-js/app/scripts/app.ts (hunk 2 above)
Comment (maffoo) [inline client-js/app/scripts/app.ts]: removed

File: client-js/app/scripts/places.ts (hunk 3 above)
Comment (maffoo) [inline client-js/app/scripts/places.ts:2]: That's correct. The "routing" happens in `app.ts` which uses the page.js library to set up routes that launch various activities. We should probably refactor this to keep the info about creating and parsing urls in this one module, but that will be for another PR.

**Comment by maffoo** [comment]
I added a few comments in `app.ts` to explain how the routes and activities are used, and how the routes relate to the URL functions in the places module. As I mentioned above, there's a bit more refactoring I want to do to keep all the URL parsing/generating code in one place instead of split between `app.ts` and `places.ts`. I think once that is in place it'll be more clear what needs to be done to add new pages to the application.

**Comment by joshm-3b6b58** [comment]
LGTM
